library/leaderboard.js

```diff
@@ -1,153 +1,69 @@
-    // find out what users are currently in permitted voice channels, then add their times as if current.
-    let guildInfo = await client.guildRepository.load(guild.id)
-    let currentPrackers = new Map()
-
-    guildInfo.permitted_channels
-      .map(chanId => guild.channels.get(chanId))
-      .filter(chan => chan != null)
-      .forEach(chan => {
-        chan.members
-          .filter(member => !member.mute && member.s_time != null && !member.deleted)
-          .forEach(member => currentPrackers.set(member.user.id, moment().unix() - member.s_time))
-      })
-
-    return currentPrackers
-  }
-
-  client.getOverallLeaderboard = async (guild) => {
-    return client.getLeaderboard(guild, p => p.overall_session_playtime,
-      size => client.userRepository.loadTopOverall(size))
+class Leaderboard {
+  constructor (repository, key) {
+    this.repository = repository
+    this.key = key
+    this.resetPage()
+    this.refresh(new Map())
   }
 
-  client.getWeeklyLeaderboard = async (guild) => {
-    return client.getLeaderboard(guild, p => p.current_session_playtime,
-      size => client.userRepository.loadTopSession(size))
-  }
-
-  client.getLeaderboard = async (guild, playtimeFn, loadTopFn) => {
-    let leaderboard = []
-
-    let topPrackers = await loadTopFn(settings.leaderboard_size)
-    topPrackers.forEach(pracker => leaderboard.push({ userId: pracker.id, time: playtimeFn(pracker) }))
-
-    let currentPrackers = await client.findCurrentPrackers(guild)
-    for (let [currentId, currentTime] of currentPrackers) {
-      let lbEntry = leaderboard.find(p => p.userId === currentId)
-      if (lbEntry == null) {
-        // we didn't get this current pracker as part of the top N, so pull from DB and calculate their total time.
-        let currentPracker = await client.userRepository.load(currentId)
-        if (currentPracker != null) {
-          currentTime += playtimeFn(currentPracker)
-        }
-
-        leaderboard.push({ userId: currentId, time: currentTime })
-      } else {
-        // this current pracker is part of the top N. Calculate their total time.
-        lbEntry.time += currentTime
+  // refresh the cache based on a pull from repository + provided liveness data
+  async refresh (liveData) {
+    this.cache = []
+    let storedData = await this.repository.loadRowsWithNonZeroKeyValue(this.key)
+    let pushed = new Set()
+    storedData.forEach(row => {
+      let totalTime = row[this.key]
+      let liveTime = liveData.get(row.id)
+      if (liveTime != null) {
+        totalTime += liveTime
+        pushed.add(row.id)
       }
-    }
 
-    leaderboard.sort((a, b) => b.time - a.time)
+      this.cache.push({ id: row.id, time: totalTime })
+    })
 
-    let msgStr = ''
-    for (let j = 0; j < settings.leaderboard_size; j++) {
-      if (leaderboard[j] == null || leaderboard[j].time === 0) {
-        break
+    liveData.forEach((key, value) => {
```

Comment:
The callback in Map.forEach() has value as its first argument and key as its second.